File: java/src/test/java/com/noodlesandwich/lists/FunctionalListTest.java
package com.noodlesandwich.lists;

import org.junit.Test;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.not;

public final class FunctionalListTest {
    @Test public void
    is_equal_to_another_list_containg_the_same_values() {
        FunctionalList<Integer> one = FunctionalList.of(1, 3, 5, 7, 9);
        FunctionalList<Integer> two = FunctionalList.of(1, 3, 5, 7, 9);
        assertThat(one, is(equalTo(two)));
    }

    @Test public void
    is_not_equal_to_another_list_containg_different_values() {
        FunctionalList<Integer> one = FunctionalList.of(1, 1, 2, 3, 5);
        FunctionalList<Integer> two = FunctionalList.of(1, 3, 5, 7, 9);
        assertThat(one, is(not(equalTo(two))));
    }

    @Test public void
    can_be_mapped_over_a_function() {
        FunctionalList<Integer> numbers = FunctionalList.of(3, 7, 10);
        assertThat(numbers.map(add(6)), is(FunctionalList.of(9, 13, 16)));
    }

    private static Function<Integer, Integer> add(final int amount) {
        return new Function<Integer, Integer>() {
            @Override public Integer apply(Integer input) {
                return input + amount;
            }
        };
    }
}

File: csharp/NoodleSandwich.Functional.Test/FunctionalListTest.cs
using System;
using NUnit.Framework;

namespace SamirTalwar.Functional
{
    [TestFixture]
    public class FunctionalListTest
    {
        [Test, ExpectedException(typeof(InvalidOperationException))]
        public void ShouldThrowExceptionIfAccessingHeadOnNil()
        {
            var head = FunctionalList<int>.Nil().Head;
        }

        [Test, ExpectedException(typeof(InvalidOperationException))]
        public void ShouldThrowExceptionIfAccessingTailOnNil()
        {
            var head = FunctionalList<int>.Nil().Tail;
        }

        [Test]
        public void ShouldBeEqualToAnotherListIfBothListsAreNil()
        {
            var one = FunctionalList<int>.Nil();
            Assert.That(one.Equals(one), Is.True);
        }

        [Test]
        public void ShouldBeEqualToAnotherListContainingTheSameValues()
        {
            var one = FunctionalList<int>.Of(1, 3, 5, 7, 9);
            var two = FunctionalList<int>.Of(1, 3, 5, 7, 9);
            Assert.That(one.Equals(two), Is.True, "should be equal if both lists contain same values");
            Assert.That(one.GetHashCode(), Is.EqualTo(two.GetHashCode()), "hash code should be equal for lists with same values");
        }

        [Test]
        public void ShouldNotBeEqualToAnObjectOfDifferentType()
        {
            var one = FunctionalList<int>.Of(1, 3, 5, 7, 9);
            Assert.That(one.Equals(new object()), Is.False, "cons should not be equal to an object of a different type");
            Assert.That(FunctionalList<int>.Nil().Equals(new object()), Is.False, "nil should not be equal to an object of a different type");
        }

        [Test]
        public void ShouldBeEqualToSelf()
        {
            var one = FunctionalList<int>.Of(1, 3, 5, 7, 9);
            Assert.That(one.Equals(one), Is.True, "cons should be equal to self");
        }

        [Test]
        public void ShouldBeNotEqualToAnotherListContainingDifferentValues()
        {
            var one = FunctionalList<int>.Of(1, 1, 2, 3, 5);
            var two = FunctionalList<int>.Of(1, 3, 5, 7, 9);
            Assert.That(one.Equals(two), Is.False, "cons should not be equal to a list with different elements");
            Assert.That(one.GetHashCode(), Is.Not.EqualTo(two.GetHashCode()), "hash code should not be equal for lists with different values");
        }

        [Test]
        public void ShouldBeNotEqualToAnotherListWhichIsNil()
        {
            var one = FunctionalList<int>.Of(1, 1, 2, 3, 5);
            var two = FunctionalList<int>.Nil();
            Assert.That(one.Equals(two), Is.False, "cons should not be equal to empty list");
            Assert.That(two.Equals(one), Is.False, "nil should not be equal to non empty list");
            Assert.That(one.GetHashCode(), Is.Not.EqualTo(two.GetHashCode()), "hash code should not be equal for lists with different values");
        }

        [Test]
        public void ShouldBeMappableOverAFunction()
        {
            var numbers = FunctionalList<int>.Of(3, 7, 10);
            Assert.That(numbers.Map(AddSixAndStringify), Is.EqualTo(FunctionalList<string>.Of("9", "13", "16")));
        }

        private string AddSixAndStringify(int input)
        {
            return (input + 6).ToString();
        }
    }
}
